### PR TITLE
[Docs][Serve] Speed up weights loading

### PR DESCRIPTION
**Pull request by cblmemo**
Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: docs/source/serving/sky-serve.rst [hunk 1]
```diff
@@ -521,3 +521,64 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
...
+  :emphasize-lines: 7
+
+    service:
+      replicas: 2
+      readiness_probe: /v1/models
+
+    resources:
+      accelerators: A100:1
+      image_id: docker:docker-image-with-model-weights
+
```
Comment (Michaelvll) [inline docs/source/serving/sky-serve.rst:583]: We should still have `run` section here, as we don't run the ENTRYPOINT of the docker image?

Comment (Michaelvll) [reply docs/source/serving/sky-serve.rst:583]: Maybe we can just recommend `docker run` in the `run` section directly?

Comment (Michaelvll) [reply docs/source/serving/sky-serve.rst:600]: Sorry, I mean instead of specifying the docker image in `image_id:`, we just ask the user to specify `docker run custom-image` in the `run` section of our yaml.

File: docs/source/serving/sky-serve.rst [hunk 2]
```diff
@@ -521,3 +521,83 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
   These settings will not take effect if you have an existing controller (either
   stopped or live).  For them to take effect, tear down the existing controller
   first, which requires all services to be terminated.
+
+Speedup Weights Loading in Large Model Serving
+----------------------------------------------
+
+When serving large models, the weights of the model are loaded from the cloud storage / public internet to the VMs. This process can take a long time, especially for large models. To speed up the weights loading, you can use the following methods:
```
Comment (Michaelvll) [inline docs/source/serving/sky-serve.rst:528]: It seems the benefit of using machine image or docker image is more for reducing the setup time, instead of the the model weight downloading time, as they are mostly for packaging the dependencies and not neccessarily means it will speed up the download of the image or the model, which should be limited by the network bandwidth instead.

Should we rewrite the section as reducing the overhead of environment setup?

File: docs/source/serving/sky-serve.rst (hunk 1 above)
Comment (cblmemo) [reply docs/source/serving/sky-serve.rst:600]: This was originally posted by zongheng [here](https://skypilot-org.slack.com/archives/C03J2KQQZSS/p1706847782816049?thread_ts=1706841328.945889&cid=C03J2KQQZSS) and actually I'm a little bit confused after a second look: will it accelerate the process? We still need to download from the docker registry and what is the difference compared to downloading from, say, hugging face..? cc @concretevitamin for a look

Comment (cblmemo) [reply docs/source/serving/sky-serve.rst:600]: Oh Zhanghao just explain that to me in the following comments. nvm :)

File: docs/source/serving/sky-serve.rst (hunk 2 above)
Comment (cblmemo) [reply docs/source/serving/sky-serve.rst:528]: I guess there are some speedups if using a machine image? That should be optimized by the cloud provider which makes it faster than a plain network download?

Comment (Michaelvll) [reply docs/source/serving/sky-serve.rst:528]: Machine images are intended to be used in a single region. It can be used in another region for launching a VM, but there will involve data transfer from one region to another. Cloud provider should have optimized it, but we probably want to be careful about the wording to avoid having a impression that all the benefits comes from weight loading.

File: docs/source/serving/sky-serve.rst [hunk 3]
```diff
@@ -521,3 +521,73 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
   These settings will not take effect if you have an existing controller (either
   stopped or live).  For them to take effect, tear down the existing controller
   first, which requires all services to be terminated.
+
+Speedup Setup in Model Serving
+------------------------------
+
+When serving AI models, the setup process like dependencies installation takes a lot of time on the cold-start time. According to one experiment we found that, dependencies installation can take 45% of times (271s out of 608s) of the setup process. To speed up this process, you can use the following methods:
```
Comment (concretevitamin) [inline docs/source/serving/sky-serve.rst:528]: Explicitly mention downloading model weights.  It's not commonly thought as being the same as dependency installation.

File: docs/source/serving/sky-serve.rst [hunk 4]
```diff
@@ -521,3 +521,73 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
...
+
+Use Machine image
```
Comment (concretevitamin) [inline docs/source/serving/sky-serve.rst:530]:
```suggestion
Use a Custom Image for VMs
```
Replace "Machine image" throughout?

"Machine image" is only GCP parlance and they  have another type of images. AWS calls it AMI.

File: docs/source/serving/sky-serve.rst [hunk 5]
```diff
@@ -521,3 +521,73 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
   These settings will not take effect if you have an existing controller (either
   stopped or live).  For them to take effect, tear down the existing controller
   first, which requires all services to be terminated.
+
+Speedup Setup in Model Serving
```
Comment (concretevitamin) [inline docs/source/serving/sky-serve.rst:525]:
```suggestion
Speeding Up Replica Setup
```

File: docs/source/serving/sky-serve.rst [hunk 6]
```diff
@@ -521,3 +521,73 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
...
+------------------------------
+
+When serving AI models, the setup process like dependencies installation and model weights downloading takes a lot of time on the cold-start time. According to one experiment we found that, dependencies installation can take 45% of times (271s out of 608s) of the setup process, and 16% for model weights downloading (98s out of 608s). To speed up this process, you can use the following methods:
```
Comment (concretevitamin) [inline docs/source/serving/sky-serve.rst:528]:
```suggestion
When serving AI models, the setup process like dependencies installation and model weights downloading may take a lot of time. To speed up this process, you can use the following methods:
```

The deleted parts do not feel like belong to docs.

File: docs/source/serving/sky-serve.rst [hunk 7]
```diff
@@ -521,3 +521,73 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
...
+
+Use a Custom Image for VMs
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can use a custom image that has the dependencies preinstalled (also probably model weights preloaded). This can be done by creating a VM, installing the dependencies, and then creating a custom image from the VM on the cloud console. Then, you can use the custom image to launch the VMs for serving:
```
Comment (concretevitamin) [inline docs/source/serving/sky-serve.rst:533]:
```suggestion
You can use a custom image that has the dependencies preinstalled and, optionally, model weights downloaded. This can be done by creating a VM, installing the dependencies, and then creating a custom image from the VM on the cloud console. Then, you can use the custom image to launch the VMs for serving:
```

File: docs/source/serving/sky-serve.rst [hunk 8]
```diff
@@ -521,3 +521,73 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
...
+
+.. code-block:: yaml
+  :emphasize-lines: 7-8
```
Comment (concretevitamin) [inline docs/source/serving/sky-serve.rst:536]:
```suggestion
  :emphasize-lines: 8-9
```

File: docs/source/serving/sky-serve.rst [hunk 9]
```diff
@@ -521,3 +521,73 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
...
+      replicas: 2
+      readiness_probe: /v1/models
+
+    resources:
+      ports: 8080
+      accelerators: A100:1
+      cloud: gcp
+      image_id: projects/my-project/global/machineImages/image-with-dependency-installed
+
+    # Here goes the setup and run commands...
```
Comment (concretevitamin) [inline docs/source/serving/sky-serve.rst:548]:
```suggestion
    # Followed by setup and run commands.
```

File: docs/source/serving/sky-serve.rst [hunk 10]
```diff
@@ -521,3 +521,73 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
...
+
+Notice that the :code:`cloud` field must be specified when :code:`image_id` is used. You can use :code:`any_of` in :code:`resources` to make it support multiple clouds:
```
Comment (concretevitamin) [inline docs/source/serving/sky-serve.rst:550]:
```suggestion
Notice that the :code:`cloud` field must be specified when :code:`image_id` is used. 

To support launching on multiple clouds, use :code:`any_of` in :code:`resources` :
```

Avoid using "you" in docs

File: docs/source/serving/sky-serve.rst [hunk 11]
```diff
@@ -521,3 +521,73 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
...
+
+.. code-block:: yaml
+  :emphasize-lines: 7-13
```
Comment (concretevitamin) [inline docs/source/serving/sky-serve.rst:553]:
```suggestion
  :emphasize-lines: 8-14
```

File: docs/source/serving/sky-serve.rst [hunk 12]
```diff
@@ -521,3 +521,73 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
...
+      accelerators: A100:1
+      any_of:
+        - cloud: gcp
+          image_id: projects/my-project/global/machineImages/image-with-dependency-installed
+        - cloud: aws
+          image_id:
+            us-east-1: ami-0729d913a335efca7
+            us-west-2: ami-050814f384259894c
+
+    # Here goes the setup and run commands...
```
Comment (concretevitamin) [inline docs/source/serving/sky-serve.rst:570]:
```suggestion
    # Followed by setup and run commands.
```

File: docs/source/serving/sky-serve.rst [hunk 13]
```diff
@@ -521,3 +521,73 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
...
+
+Notice that AWS needs a per-region image since its image is only valid in a single region. This requires to configure a lot of images on the cloud console, but it can significantly speed up the dependency installation. It also takes the advantage of the in-region fast network, which delivers a quicker download speed of the machine image.
```
Comment (concretevitamin) [inline docs/source/serving/sky-serve.rst:572]:
```suggestion
Notice that AWS needs a per-region image since an AMI image is associated with a particular region. While this requires creating several images on the cloud console, it can significantly speed up the dependency installation. This approach also takes advantage of the in-region fast network, which delivers a quicker download speed of the machine image.
```

File: docs/source/serving/sky-serve.rst [hunk 14]
```diff
@@ -521,3 +521,73 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
...
+
+Use Docker Container as Runtime Environment
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (concretevitamin) [inline docs/source/serving/sky-serve.rst:575]:
```suggestion
Use Docker Containers as Runtime Environment
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: docs/source/serving/sky-serve.rst [hunk 15]
```diff
@@ -521,3 +521,73 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
...
+
+You can also :ref:`use docker containers as runtime environment <docker-containers-as-runtime-environments>`. This can be done by creating a Docker container that has the dependencies preinstalled, and then using the container to launch the VMs for serving:
```
Comment (concretevitamin) [inline docs/source/serving/sky-serve.rst:577]:
```suggestion
You can also :ref:`use docker containers as runtime environment <docker-containers-as-runtime-environments>`. This can be done by creating a Docker image that has the dependencies preinstalled and (optionally) model weights downloaded, and then using the container image to launch the VMs for serving:
```

File: docs/source/serving/sky-serve.rst [hunk 16]
```diff
@@ -521,3 +521,73 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
...
+
+.. code-block:: yaml
+  :emphasize-lines: 7
```
Comment (concretevitamin) [inline docs/source/serving/sky-serve.rst:580]:
```suggestion
  :emphasize-lines: 8
```

File: docs/source/serving/sky-serve.rst [hunk 17]
```diff
@@ -521,3 +521,73 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
...
+    service:
+      replicas: 2
+      readiness_probe: /v1/models
+
+    resources:
+      ports: 8080
+      accelerators: A100:1
+      image_id: docker:docker-image-with-dependency-installed
+
+    # Here goes the setup and run commands...
```
Comment (concretevitamin) [inline docs/source/serving/sky-serve.rst:591]: ditto

File: docs/source/serving/sky-serve.rst [hunk 18]
```diff
@@ -521,3 +521,73 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
...
+
+This is easier to configure than machine images, but it may have a longer startup time than machine images since it needs to pull the docker image from the registry.
```
Comment (concretevitamin) [inline docs/source/serving/sky-serve.rst:593]:
```suggestion
This is easier to configure than machine images, but it may have a longer startup time than machine images since it needs to pull the docker image from the registry in addition to the usual VM setup.
```

Comment (concretevitamin) [inline docs/source/serving/sky-serve.rst:593]: Have we actually timed these two methods?

Comment (cblmemo) [reply docs/source/serving/sky-serve.rst]: I'm feeling like it is hard to make a fair comparison - it is largely dependent on the base docker/machine image used... Though I'll try to make some benchmarks and see the results 🫡

File: docs/source/serving/sky-serve.rst [hunk 19]
```diff
@@ -521,3 +521,75 @@ The :code:`resources` field has the same spec as a normal SkyPilot job; see `her
...
+      replicas: 2
+      readiness_probe: /v1/models
+
+    resources:
+      ports: 8080
+      accelerators: A100:1
+      image_id: docker:docker-image-with-dependency-installed
+
+    # Followed by setup and run commands.
+
```
Comment (Michaelvll) [inline docs/source/serving/sky-serve.rst:594]: We could mention something about how this docker image should be built, especially, it could have SkyPilot runtime pre-built. Something like the following would be useful (could you help giving an concrete example for how to install vllm, and download the image in the Dockerfile below for a better reference, i.e. replacing the line `# Your dependencies installation and model download code goes here` with actual workable commands for serving vllm+mistral):

``````
Your docker image can have all skypilot dependencies pre-installed to further reduce the setup time, you could try building your docker image based from our base image. The `Dockerfile` could look like the following:
```Dockerfile
FROM docker:berkeleyskypilot/skypilot-k8s-gpu:latest

# Your dependencies installation and model download code goes here
```
``````

**Review by Michaelvll — COMMENTED** [review]
Thanks for adding this doc @cblmemo! We have users asking for this and it would be nice we can directly point them to this page. : )

**Comment by cblmemo** [comment]
> Thanks for adding this doc @cblmemo! We have users asking for this and it would be nice we can directly point them to this page. : )

Just want to update the status of this PR first: I found a mysterious bug that causes an NVML initialization error when using docker container as runtime env. By bisect it seems like those lines are causing the error: https://github.com/skypilot-org/skypilot/blob/82c50f5c281fc88eb417dfd928dbe570da99aaf4/sky/provision/docker_utils.py#L271-L274

That is very strange since those are running on the host but somehow affect the containers. Will investigate more.

**Comment by Michaelvll** [comment]
We should consider having this PR updated and merged as well. : )

**Comment by cblmemo** [comment]
> We should consider having this PR updated and merged as well. : )

This PR is blocked by the max/ultra disk tier as the current performance is not better than install everything from pip...

**Comment by Michaelvll** [comment]
Another user requests this. : )

**Comment by cblmemo** [comment]
Left some benchmark results using PR #3860. In the following table, `high` = gp3 7,000 IOPS, `ultra` = io2 20,000 IOPS, `max` = io2 100,000 IOPS. All tests are running on AWS and the result is the e2e execution time for launching a Llama 2 70b checkpoint w/ the latest version of vLLM, on an `A10G:8` instance.

||`high`|`ultra`|`max`|
|-|-|-|-|
|Use AMI|> 2 hours|487s|467s|
|Download from HF|524s|410s|-|

In conclusion, our `high` disk tier is indeed not enough for large checkpoint downloading and the `ultra` tier increases the performance a lot. Though the AMI does not enhance the performance as expected; there might be other bottlenecks like the download speed of the AMI.

**Comment by Michaelvll** [comment]
We should revamp this PR with our latest findings and support for ultra disk : )

**Comment by cblmemo** [comment]
> We should revamp this PR with our latest findings and support for ultra disk : )

Done in #3949 . Still keeping this so we could investigate if it is possible to speed up by using AMI & docker image.
